server/plugin.go

```diff
@@ -599,22 +598,51 @@ func (p *Plugin) HasGroupHook(ctx context.Context, user *gitlab.UserInfo, namesp
 	return found, err
 }
 
-func (p *Plugin) checkAndRefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
+func (p *Plugin) checkAndRefreshToken(userInfo *gitlab.UserInfo) error {
 	// If there is only one minute left for the token to expire, we are refreshing the token.
 	// The detailed reason for this can be found here: https://github.com/golang/oauth2/issues/84#issuecomment-831492464
 	// We don't want the token to expire between the time when we decide that the old token is valid
 	// and the time at which we create the request. We are handling that by not letting the token expire.
-	if time.Until(token.Expiry) <= 1*time.Minute {
+
+	if time.Until(userInfo.Token.Expiry) <= 1*time.Minute {
```

Comment:
Also, we need to load the token from the kv store while we have the mutex locked, otherwise we aren't synchronizing access to the token.

Reply:
The mutex is in `attachUserContext` because I wanted to lock it right before calling `getGitlabUserInfoByMattermostID`.

Reply:
I'm not sure how it will work to move the mutex to the `checkAndRefreshToken` function, as I need to lock it before calling `GetGitlabUserInfoByMattermostId`, and only unlock it after the logic in `checkAndRefreshToken`, right?

If you're suggesting reverting the `checkToken` function I added most recently and going back to having more per-function calls to `checkAndRefreshToken`, that's fine, but I'm not sure about the mutex.

Can you lay out how you see it working for one request, where do you see each of: acquiring the mutex lock, loading the token from KV, check and refresh and save to KV, and unlock mutex?

Reply:
Are you suggesting going closer to the original solution implemented in the previous token fix PR (the current `master` code), where `checkAndRefreshToken` is just called from `GetGitlabUserInfoByMattermostId`? We could do that, coding a workaround for the loop issue I was having (described earlier in the thread) would be simple enough given how much work it's ended up to do it another way. Something like a function that loads the user info without checking the token (since that was causing the loop) which could be used by `p.DisconnectGitlabAccount`.